tensorflow/lite/micro/kernels/softmax.cc

```diff
@@ -72,12 +99,46 @@ TfLiteStatus CalculateSoftmaxParams(TfLiteContext* context,
 
 }  // namespace
 
+void* SoftmaxInit(TfLiteContext* context, const char* buffer, size_t length) {
+  void* raw = nullptr;
+  context->AllocatePersistentBuffer(context, sizeof(SoftmaxOpData), &raw);
+  return raw;
+}
+
 TfLiteStatus SoftmaxPrepare(TfLiteContext* context, TfLiteNode* node) {
   TF_LITE_ENSURE_EQ(context, NumInputs(node), 1);
   TF_LITE_ENSURE_EQ(context, NumOutputs(node), 1);
   const TfLiteTensor* input = GetInput(context, node, 0);
   TF_LITE_ENSURE(context, NumDimensions(input) >= 1);
 
+  auto* params = reinterpret_cast<TfLiteSoftmaxParams*>(node->builtin_data);
+  SoftmaxOpData* data = new (node->user_data) SoftmaxOpData;
+
+  TfLiteTensor* output = GetOutput(context, node, 0);
+  if (output->type == kTfLiteInt16) {
+    TF_LITE_ENSURE(context, input->type == kTfLiteInt8 ||
+                                input->type == kTfLiteUInt8 ||
+                                input->type == kTfLiteInt16);
+  } else {
+    TF_LITE_ENSURE_EQ(context, input->type, output->type);
+  }
+
+  // Populate LUT if required
+  if (input->type == kTfLiteInt16) {
+    TF_LITE_ENSURE_EQ(context, output->params.zero_point, 0);
+
+    data->params.exp_lut = data->exp_lut;
+    // exp LUT only used on negative values
+    // we consider exp(-10.0) is insignificant to accumulation
+    gen_lut([](double value) { return std::exp(value); }, -10.0, 0.0,
```

Comment:
Do these have to be doubles, or would floats work? Most of our platforms will fall back to software emulation for doubles, which can be slow, even just for initializations.

Reply:
Done changed to float.